src/NzbDrone.Core/Download/Clients/rTorrent/RTorrent.cs

```diff
@@ -0,0 +1,217 @@
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Threading;
+using NzbDrone.Common.Disk;
+using NzbDrone.Common.Extensions;
+using NzbDrone.Common.Http;
+using NzbDrone.Core.Configuration;
+using NzbDrone.Core.MediaFiles.TorrentInfo;
+using NLog;
+using NzbDrone.Core.Validation;
+using FluentValidation.Results;
+using System.Net;
+using NzbDrone.Core.Parser.Model;
+using NzbDrone.Core.RemotePathMappings;
+
+namespace NzbDrone.Core.Download.Clients.RTorrent
+{
+    public class RTorrent : TorrentClientBase<RTorrentSettings>
+    {
+        private readonly IRTorrentProxy _proxy;
+
+        public RTorrent(IRTorrentProxy proxy,
+                        ITorrentFileInfoReader torrentFileInfoReader,
+                        IHttpClient httpClient,
+                        IConfigService configService,
+                        IDiskProvider diskProvider,
+                        IRemotePathMappingService remotePathMappingService,
+                        Logger logger)
+            : base(torrentFileInfoReader, httpClient, configService, diskProvider, remotePathMappingService, logger)
+        {
+            _proxy = proxy;
+        }
+
+        protected override string AddFromMagnetLink(RemoteEpisode remoteEpisode, string hash, string magnetLink)
+        {
+            _proxy.AddTorrentFromUrl(magnetLink, Settings);
+
+            // Wait until url has been resolved before returning
+            var TRIES = 5;
+            var remainingTries = TRIES;
+            var RETRY_DELAY = 500; //ms
+            var ready = false;
+            while (true)
```

Comment:
> If hash has torrent, it will sleep unnecessary

No, it will break immediately.

> Variable "ready" is used afterwards, to avoid making this loop too deep

Right, then a two liner for it makes sense.